### ISSUE TITLE
Question: extracting alignment from profiles

### ISSUE THREAD

**Issue by acpguedes** [soedinglab/MMseqs2#153]
Hi, just a simple question, since I have profiles build with result2profile, how do I extract the alignment from each profile?
Stockholm, fasta, a2m, a3m, or any well known format. 

Also, a related question, may I use the profiles directly to hhalign? Yet, there are huge difference between compare profile with mmseqs result2profile and use hhalign?

**Comment by milot-mirdita** [comment]
The `result2msa` module will generate (center star) MSAs from an alignment result. `convertalis` can print the pairwise alignments. Its not possible to create alignments from profiles, since the required alignment information is already lost at that point.

We don't support the hhmake profile format in MMseqs2, you would have to create the MSAs and build new profiles using `hhmake`.  In the Uniclust databases we use ClustalO to create MSAs for each MMseqs2 cluster, which are then converted to the A3M alignment format with `reformat.pl` and hhm profile format through `hhmake`:

https://github.com/soedinglab/uniclust-pipeline/blob/develop/hhdatabase/make_hhdatabase.sh

**Comment by acpguedes** [comment]
Thanks for fast answer.

**Comment by acpguedes** [comment]
I used `mmseqs search PROFILEDB /databases/fadb/freeze/all.mmseqs pbp.found_with_profile tmp --threads 10 -a -c 0.8` follow by `mmseqs result2msa PROFILEDB /databases/fadb/freeze/all.mmseqs pbp.found_with_profile.MSADB pbp.found_with_profile.msa` and `mmseqs convert2fasta pbp.found_with_profile.ALIGNEDREGIONS pbp.found_with_profile.ALIGNEDREGIONS.fa`. But this returned me sequences with query header also repeated a lot of times.

Actually, convertalis crash with me, seams that it print a TSV that could include CIGAR alignment, but I didn't understand how to use it to print alignments.

Could you be just a little more specific in how do I extract alignments from cluster DB and also from search DB if possible? 
Both them should pass by result2msa? How does each result's set will be printed?

**Comment by milot-mirdita** [comment]
Could you provide the call and log and if possible also a backtrace (see https://github.com/soedinglab/MMseqs2/issues/136#issuecomment-441411023) of the `convertalis` crash? For the pairwise alignments the `--format-output` parameter needs to include `qaln,taln`.

Convert2fasta turns a sequence into a flat FASTA file. Its not useful for result databases. 

Search databases should work directly (please provide the log so we can see what went wrong). For cluster databases you have to align the result clustering with the `align -a` module and parameter to generate the alignment information and backtrace and then call `convertalis`.

**Comment by acpguedes** [comment]
Thanks a lot for the help, @milot-mirdita 

Build PROFILEDB and search with this was successful.
final steps:
> mmseqs search PROFILEDB /databases/fadb/freeze/all.mmseqs pbp.found_with_profile tmp --threads 10 -a -c 0.8 --cov-mode 2 --add-self-matches 
> mmseqs convertalis PROFILEDB /databases/fadb/freeze/all.mmseqs pbp.found_with_profile pbp.found_with_profile.tsv   

At this point, everything okay.  
The step below also was performed nicely without errors.  
> mmseqs extractalignedregion PROFILEDB /databases/fadb/freeze/all.mmseqs pbp.found_with_profile pbp.found_with_profile.ALIGNEDREGION  

The crash happens here:
> mmseqs convert2fasta pbp.found_with_profile.ALIGNEDREGIONS pbp.found_with_profile.ALIGNEDREGIONS.fa   

With this message:
```

> Program call:
> convert2fasta pbp.found_with_profile.ALIGNEDREGIONS pbp.found_with_profile.ALIGNEDREGIONS.fa 
> 
> MMseqs Version:   	7-4e23d
> Use ffindex header	false
> Verbosity         	3
> 
> Could not open data file pbp.found_with_profile.ALIGNEDREGIONS!

```

But I realised maybe I did everything wrong.  

What I want, actually, is the MSA for each cluster with only aligned regions.  
Also, I want the MSA of aligned regions from a search result.

The deal is: I have 501 queries single domain, I would to use this against a 17M sequences DB, since the matches probably have at least 1 region contains the same domain as the query, I would like the MSA or even a sequence DB itself contain **only the regions of the match**.  
Extracting only the region of the matches allow me to infer more accurate alignments and phylogeny. Also allow me build specifics models.  

Could you provide me a simple workflow to this task?

Thanks in advance.

**Comment by milot-mirdita** [comment]
I am not sure if that makes sense. Our alignments are always local and furthermore `result2msa` will never generate gap columns in the query(/representative) sequence.  So the MSAs will never contain unaligned regions.

You should also not add `--add-self-matches` that option only makes sense with clustering, where the query database is the same as the target database.

```
mmseqs search PROFILEDB /databases/fadb/freeze/all.mmseqs pbp.found_with_profile tmp --threads 10 -a -c 0.8 --cov-mode 2 
mmseqs result2msa PROFILEDB /databases/fadb/freeze/all.mmseqs pbp.found_with_profile msa --threads 10
```

This should already do what you want without adding further complicated steps. For downstream analysis you can use the `apply` module that will call an external program on each individual entry (passing the entry to stdin and writing the result from stdout to a new database). Or you could use the `ffindex_unpack` binary from the hh-suite to create separate files for each database entry.

**Comment by acpguedes** [comment]
Thanks. This works fine.

> mmseqs result2msa pbps.DB pbps.DB pbps.mergedDB pbps.mergedDB.msa

I used 0 of coverage and 20% of identity to cluster results.  
By checking the alignments, I realize that some sequences only keep a small region. I also saw the command `mmseqs expandaln` which makes me wonder if it can be used to extend sequences where alignment is filled with gaps.

Since `mmseqs expandaln` require 4 DBs what is the fourth `<i:resultDB|ca3mDB>`.
